### PR TITLE
Support SHA hashing

### PR DESCRIPTION
**Pull request by haberdashPI**

### CODE REVIEW

File: test/runtests.jl [hunk 1]
```diff
@@ -47,18 +48,23 @@ StableHashTraits.hash_method(::TestType, ::MyContext) = UseQualifiedName(UseProp
     @test stable_hash([1, 2, 3]) == 0x1a366aea
     @test stable_hash((a=1, b=2)) == 0x240bb84c
     @test stable_hash(sin) == 0x7706a39f
-    @test stable_hash(TestType2(1, 2)) == 0x1f99ed3b
+    @test stable_hash(TestType2(1, 2)) == 0x8c28b28a
```
Comment (ericphanson) [inline test/runtests.jl:51]: this is a breaking change to the package, btw

Comment (haberdashPI) [reply test/runtests.jl:51]: Ah yes, I will at least mark it as such.

Do you think that is sufficient? I think with a little more work and a few extra methods specific to `crc32c` I could probably make sure these remain unchanged.

File: src/StableHashTraits.jl [hunk 2]
```diff
@@ -24,17 +58,18 @@ write(io, x) = Base.write(io, x)
...
 
 struct UseIterate end
 function stable_hash_helper(x, hash, context, ::UseIterate)
-    result = hash(UInt8[])
     for el in x
-        val = stable_hash_helper(el, hash, context, hash_method(el, context))
-        result = hash(copy(reinterpret(UInt8, [val])), result)
+        val = stable_hash_helper(el, similar_hasher(hash), context,
+                                 hash_method(el, context))
+        update!(hash, copy(reinterpret(UInt8, vcat(digest!(val)))))
```
Comment (kolia) [inline src/StableHashTraits.jl:70]: I'm not understanding the `vcat` here.

**Review by kolia — APPROVED** [review]
lgtm

I'm curious, what was it that was breaking hash values as Eric pointed out, and what was the fix?

File: src/StableHashTraits.jl (hunk 2 above)
Comment (haberdashPI) [reply src/StableHashTraits.jl:70]: It ensures the object is an array; `vcat(1) == [1]`; `vcat([1]) == [1]`

File: src/StableHashTraits.jl [hunk 3]
```diff
@@ -3,6 +3,40 @@ module StableHashTraits
 export stable_hash, UseWrite, UseIterate, UseProperties, UseQualifiedName
 
 using CRC32c, TupleTools, Compat, UUIDs, Dates
+using SHA: SHA
+
+#####
+##### Hash Function API 
+#####
+
+# SHA functions need to `update!` an context object for each object ot hash and then
```
Comment (hannahilea) [inline src/StableHashTraits.jl:12]:
```suggestion
# SHA functions need to `update!` an context object for each object to hash and then
```

**Comment by haberdashPI** [comment]
> I'm curious, what was it that was breaking hash values as Eric pointed out, and what was the fix?

I was handling empty lists/tuples differently in the new version. It just required adding a special case for this situation.

File: src/StableHashTraits.jl (hunk 3 above)
Comment (haberdashPI) [reply src/StableHashTraits.jl]: Thanks for the catch!

File: src/StableHashTraits.jl [hunk 4]
```diff
@@ -24,17 +58,33 @@ write(io, x) = Base.write(io, x)
 function stable_hash_helper(x, hash, context, ::UseWrite)
     io = IOBuffer()
     write(io, x, context)
-    return hash(take!(io))
+    update!(hash, take!(io))
+    return hash
+end
+
+function recursive_hash!(hash, result)
```
Comment (kolia) [inline src/StableHashTraits.jl:65]: Can you comment on when `digest!` returns `nothing`?

**Pull request merged**